plugins/logitech-tap/fu-logitech-tap-sensor-device.c

```diff
@@ -0,0 +1,371 @@
+const guchar kHidMcuSerialNumberSetReportByte4 = 0x00;
+
+/* reboot device */
+const guchar kHidReportIdMcuSetCmd = 0x1a;
+
+const guint kLogiDefaultSensorSleepIntervalMs = 50;
+
+struct _FuLogitechTapSensorDevice {
+	FuLogitechTapDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuLogitechTapSensorDevice, fu_logitech_tap_sensor_device, FU_TYPE_LOGITECH_TAP_DEVICE)
+
+static gboolean
+fu_logitech_tap_sensor_device_set_feature(FuLogitechTapSensorDevice *self, const guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	fu_dump_raw(G_LOG_DOMAIN, "HidSetFeature", data, datasz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    HIDIOCSFEATURE(datasz),
+				    (guint8 *)data,
+				    NULL,
+				    FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				    error);
+#else
+    /* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_get_feature(FuLogitechTapSensorDevice *self, guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	g_autoptr(GError) error_local = NULL;
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureReq", data, datasz);
+
+	/* try HIDIOCGINPUT request in case of failure */
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(datasz),
+				  data,
+				  NULL,
+				  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_debug("failed to send get request, retrying: %s", error_local->message);
+		if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGINPUT(datasz),
+				  data,
+				  NULL,
+				  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				  error))
+			return FALSE;
+	}
+
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureRes", data, datasz);
+
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_set_tde(FuDevice *device, const guchar tde_mode, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuTdeReportId, kHidMcuTdeModeSelector, tde_mode, 0, 0};
+
+	/* enable/disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error))
+	return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_logitech_tap_sensor_device_reboot_device(FuDevice *device,
+				   GError **error)
+{
+    FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	gint8 pinclr = 05;
+ 	guint8 pinset = 06;
+ 	guint8 PWR = 45;
+ 	guint8 RST = 46;
+ 	guint8 set_data[HID_SET_DATA_LEN] = {kHidReportIdMcuSetCmd, pinclr, PWR, 0, 0};
+    g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+
+	g_debug("trigger device reboot");
```

Comment:
given there is already going to be daemon debugging for the first command (attach) is this message really needed?

Reply:
Removed